File: lib/escher/escher/resources/resource.cc
// Copyright 2017 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "escher/resources/resource.h"

#include "escher/impl/command_buffer.h"
#include "escher/resources/resource_manager.h"

namespace escher {

const ResourceTypeInfo Resource2::kTypeInfo("Resource",
                                            ResourceType::kResource);

Resource2::Resource2(ResourceManager* owner) {
  owner->BecomeOwnerOf(this);
}

void Resource2::KeepAlive(impl::CommandBuffer* command_buffer) {
  auto sequence_number = command_buffer->sequence_number();
  FTL_DCHECK(sequence_number_ <= sequence_number);
  sequence_number_ = sequence_number;
}

const VulkanContext& Resource2::vulkan_context() const {
  FTL_DCHECK(owner());
  return owner()->vulkan_context();
}

ResourceManager* Resource2::owner() const {
  return static_cast<ResourceManager*>(
      Ownable<Resource2, ResourceTypeInfo>::owner());
}

}  // namespace escher

File: src/sys/pkg/lib/fuchsia-repo/src/resource.rs
// Copyright 2022 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use {
    crate::{range::ContentRange, repository::Error, util::read_stream_to_end},
    bytes::Bytes,
    fidl_fuchsia_pkg_ext::RepositoryConfig,
    futures::{
        future::ready,
        stream::{once, BoxStream, StreamExt},
    },
    std::{convert::TryFrom, io},
};

/// [Resource] represents some resource as a stream of [Bytes] as provided from
/// a repository server.
pub struct Resource {
    /// The range in bytes available for this resource.
    pub content_range: ContentRange,

    /// A stream of bytes representing the resource.
    pub stream: BoxStream<'static, io::Result<Bytes>>,
}

impl Resource {
    /// The length of the content in bytes in the stream. This may be smaller than the total length
    /// of the file.
    pub fn content_len(&self) -> u64 {
        self.content_range.content_len()
    }

    /// The total length of the file range in bytes. This may be larger than the bytes in the stream.
    pub fn total_len(&self) -> u64 {
        self.content_range.total_len()
    }

    pub async fn read_to_end(&mut self, buf: &mut Vec<u8>) -> Result<(), Error> {
        buf.reserve(self.content_len() as usize);
        read_stream_to_end(&mut self.stream, buf).await.map_err(Error::Io)
    }
}

impl std::fmt::Debug for Resource {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("resource")
            .field("content_range", &self.content_range)
            .field("stream", &"..")
            .finish()
    }
}

impl TryFrom<RepositoryConfig> for Resource {
    type Error = Error;

    fn try_from(config: RepositoryConfig) -> Result<Resource, Error> {
        let json = Bytes::from(serde_json::to_vec(&config).map_err(|e| anyhow::anyhow!(e))?);
        let complete_len = json.len() as u64;
        Ok(Resource {
            content_range: ContentRange::Full { complete_len },
            stream: once(ready(Ok(json))).boxed(),
        })
    }
}
